### PR TITLE
Emit events for errors in the namespace controller to help understand stuck deletion

### PR DESCRIPTION
**Pull request by alvaroaleman**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

/kind bug
/sig apimachinery

/assign @lavalamp @wozniakjan 

**What this PR does / why we need it**:

Makes the namespace controller emit events on errors to help humans understand what the issue is. Also includes the GVR into the error that is used for the event.

Done in order to help with https://github.com/kubernetes/kubernetes/issues/70916

With this change, a namespace that is stuck in deletion looks like this:

```
$ /tmp/kubectl describe ns kubevirt
Name:         kubevirt
Labels:       kubevirt.io=
Annotations:  kubectl.kubernetes.io/last-applied-configuration:
                {"apiVersion":"v1","kind":"Namespace","metadata":{"annotations":{},"labels":{"kubevirt.io":""},"name":"kubevirt"}}
Status:       Terminating

No resource quota.

No LimitRange resource.
Events:
  Type     Reason                   Age                   From                          Message
  ----     ------                   ----                  ----                          -------
  Warning  DeletionFailed           3s (x4 over 27s)      namespace_controller          some content ([kubevirt.io/v1alpha3, Resource=kubevirts]) remains in the namespace, estimate 15 seconds before it is removed
```

This is in my opinion for more useful than the Conditions we have today.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
The namespace controller now emits events when it encounters an error, helping to understand why namespace get stuck in deletion
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
none
```

### CODE REVIEW

**Comment by liggitt** [comment]
I'm not in favor of making these events. If the conditions on the namespace are lacking, it would make more sense to me to improve those.

**Comment by alvaroaleman** [comment]
> I'm not in favor of making these events. If the conditions on the namespace are lacking, it would make more sense to me to improve those.

What about dumping the errors into an "Error Condition"? That would hide previous errors but I guess we most likely do not care about them anyways.

Also IMHO kubectl also needs enhancing to make its `describe` include the conditions for namespaces.

**Comment by liggitt** [comment]
> What about dumping the errors into an "Error Condition"?

those are already intended to be reported via the calls to ProcessDiscoverResourcesErr/ProcessGroupVersionErr/ProcessDeleteContentErr... if there are other relevant errors, reporting those would make sense to me as well

**Comment by liggitt** [comment]
> kubectl also needs enhancing to make its `describe` include the conditions for namespaces.

that seems reasonable

**Comment by lavalamp** [comment]
I would not expect this to work, it looks to be adding events to a namespace that's in the deletion flow? Even if apiserver doesn't fail that, the namespace deleter will just delete all those events next time it runs. Additionally, even if another namespace can be found to hold these events, I'm uncomfortable with adding this as an event source, I'm not sure how many events this might generate; I don't want deleting a huge namespace (e.g. due to a runaway controller) to be a risky action.

IOW I agree w/ liggitt@; please improve the conditions and/or kubectl.

File: pkg/controller/namespace/namespace_controller.go [hunk 1]
```diff
@@ -48,6 +52,8 @@ const (
 	// * non-leader etcd servers to observe last-minute object creations in a namespace
 	//   so this controller's cleanup can actually clean up all objects
 	namespaceDeletionGracePeriod = 5 * time.Second
+	// controllerName is the name of this controller
+	controllerName = "namespace_controller"
```
Comment (lavalamp) [inline pkg/controller/namespace/namespace_controller.go]: specifically this is the namespace lifecycle controller, BTW

**Comment by alvaroaleman** [comment]
> IOW I agree w/ liggitt@; please improve the conditions and/or kubectl.

Is a kubectl change for this a "just do" thing or does it require a KEP?

**Comment by liggitt** [comment]
> > IOW I agree w/ liggitt@; please improve the conditions and/or kubectl.
> 
> Is a kubectl change for this a "just do" thing or does it require a KEP?

It's fine to just add. The human-readable `describe` output is not an API, and surfacing information added to the API is reasonable.

**Comment by liggitt** [comment]
closing this in favor of improving the reported conditions

/close

**Pull request closed without merging**